File: 733-flood-fill/flood-fill.cpp
class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        int icolor = image[sr][sc];
        if(icolor != color) {
            dfs(image, sr, sc, icolor, color);
        }
        return image;
    }

    void dfs(vector<vector<int>>& image, int row, int col, int icolor, int color){
        if(row >= 0 && row < image.size() && col >= 0 && col < image[0].size() && image[row][col] != color && image[row][col] == icolor){
            image[row][col] = color;
            dfs(image, row-1, col, icolor, color);
            dfs(image, row+1, col, icolor, color);
            dfs(image, row, col+1, icolor, color);
            dfs(image, row, col-1, icolor, color);
        }
    }
};

File: 733-flood-fill/flood-fill.swift
class Solution {
    func dfs(_ grid: inout [[Int]], _ i: Int, _ j: Int, _ start: Int, _ color: Int){
        var m = grid.count
        var n = grid[0].count
        // check if oob or cell is color or cell is start
        if(i < 0 || j < 0 || i >= m || j >= n || grid[i][j] == color || grid[i][j] != start) { return }
        grid[i][j] = color
        dfs(&grid, i, j+1, start, color)
        dfs(&grid, i+1, j, start, color)
        dfs(&grid, i, j-1, start, color)
        dfs(&grid, i-1, j, start, color)
        return
    }

    func floodFill(_ image: [[Int]], _ sr: Int, _ sc: Int, _ color: Int) -> [[Int]] {
        var image = image
        dfs(&image, sr, sc, image[sr][sc], color)
        return image
    }
}